### ISSUE TITLE
Natural Language Understanding in RapidPro

### ISSUE THREAD

**Issue by johncordeiro** [rapidpro/rapidpro#832]
I would like to present RapidPro contributors the functionality of Natural Language Processing inside its flows and triggers, to enable a more fluent conversation. This improves the usage of RapidPro in the creation of smart Chatbots.

First of all, Bothub is an [open source](https://github.com/ilhasoft/bothub) platform developed by Ilhasoft and [supported](http://unicefstories.org/2018/08/16/fundgraduate-ilhasoft/) by Unicef Innovation Fund. It was launched 3 months ago and it already supports 11 languages, the goal is to support 33 languages until the end of 2018.

Bothub's main functionalities are:

- Easy creation of the contextual NLP dataset with a few sentences and start the prediction of textual data by collecting intentions and features from it. For example:
  ![captura de tela 2018-12-07 as 09 14 01](https://user-images.githubusercontent.com/6073075/49647543-9f572580-fa01-11e8-8ab1-082bc6a074d7.png)
- The possibility of sharing NLP Datasets with other users, enabling knowledge sharing and contribution in translation and testing;
- An algorithm that enables the quick creation of a vocabulary of languages that aren't provided by other market players, democratizing the usage of NLP Datasets and delivering good accuracy to these users.

------

When it comes to RapidPro, I would like to propose the submission of a Pull Request for the full-integration with Bothub that has been continuously validated during the last 3 months in a RapidPro hosting.

The most important reason to propose the built-in capability of Bothub NLP in RapidPro comes from after some years of experience creating smart Chatbots on the platform by using webhooks to call other NLP platforms to create a fluent conversation. It has been a big problem to provide contextual understanding and also to maintain these flows. Furthermore, it's currently not possible to change the subject inside the conversation as we can do in this proposed solution.

After 3 months of usage in real cases (https://push.al/conheca-a-susana), RapidPRO + Bothub has proven to be a perfect combination that delivers one of the most powerful tools for agent creation in the world. The possibilities are endless.

The integration is basically divided into 3 steps, 1. the registration of knowledge bases into the RapidPro workspaces; 2. the creation of triggers based on Intentions retrieved by Natural Language Understanding; and 3. usage of these intentions and features from sentences inside RapidPro flows.

### 1. Registration of Bots

- In the RapidPro workspace home page, you have one more option to connect your Bothub repository:
  <img height="200" alt="screen shot 2018-12-06 at 5 36 49 pm" src="https://user-images.githubusercontent.com/6073075/49610267-dcc59f80-f97c-11e8-8ce7-d6a2f5b171e8.jpg">
- Users can add just one or more repositories in the RapidPro workspace, enabling the merging of open collection of datasets available in Bothub:
  <img width="1231" alt="screen shot 2018-12-06 at 5 36 49 pm" src="https://user-images.githubusercontent.com/6073075/49610629-edc2e080-f97d-11e8-82cd-4bf8d98bef8f.png">

### 2. Utilization in Triggers

- Once repositories are registered in the workspace, one more trigger type is enabled:
  <img width="500" alt="screen shot 2018-12-06 at 5 36 49 pm" src="https://user-images.githubusercontent.com/6073075/49611242-b5240680-f97f-11e8-8a87-c4f94ea68270.png">
- User can select the flow to start, restrict by contact groups, filter just intentions with a confidence rate of the prediction and finally the Intents that can start this flow:
  <img width="500" alt="screen shot 2018-12-06 at 5 57 42 pm" src="https://user-images.githubusercontent.com/6073075/49611490-73479000-f980-11e8-9c87-3a5dc2b5b5ee.png">

### 3. Utilization in Flows

- Uses can also make contextual predictions in Messaging Flows, on top of `Wait for response` ruleset, there is a new rule called `has an intent` and you can freely combine it with other rules:
  <img width="500" alt="screen shot 2018-12-06 at 6 03 42 pm" src="https://user-images.githubusercontent.com/6073075/49611787-6d05e380-f981-11e8-9796-3630fe15de61.png">
- Using the prediction results is also possible with flow variables, this is really useful for entities (features) inside the phrase:
  <img width="500" alt="screen shot 2018-12-06 at 6 04 40 pm" src="https://user-images.githubusercontent.com/6073075/49611826-8f97fc80-f981-11e8-9a07-77c58d2fc22c.png">
- It's also possible to ignore NLU triggers inside flows as we usually do for keyword triggers. It means that we can change the subject or not to start a different flow:
  <img width="500" alt="screen shot 2018-12-06 at 6 18 01 pm" src="https://user-images.githubusercontent.com/6073075/49647926-cfeb8f00-fa02-11e8-918a-66ba091728f0.png">

We're excited to be on the position of contributing in such a meaningful way to RapidPro with
a tool that's enabling Ilhasoft to create wonderful projects.

**Comment by nicpottier** [comment]
Very cool, have lots of thoughts on this but a few things before the weekend.

General comments:
 * We probably want to make this a generic RapidPro<->NLP Provider rather than anything specific to bothub. I would suggest we build this out with both bothub and https://wit.ai as another test case seeing as they support 50 languages and have a very open model. So we would envision that NLP providers become a bit like channel providers where you can pick who you want to use for the intent actions in your triggers and flows. (others that are worth looking at to validate our approach would be https://www.luis.ai/ and https://dialogflow.com/) Good news is the interfaces here are super simple so launching with support for a few different providers should be easy.
 * We won't fold this into the old engine at this point as we are so close to shipping the new engine on everything but we definitely are interested in adding generic NLP support to the new engine and this is a very useful starting point in designing that. It feels a bit like direct cash transfers in that it could be an extension that is optional in the engine but we can talk through various options. Important bit here is that the flow definition etc, need to all be agnostic to provider. (IE, something like a `classify` action in the flow that determines which NLP provider to use based on org configs)
 * I really like both the trigger integration and the flow integration. One missing bit may be letting the user pick a specific app that they are classifying against. Maybe NLP providers are even more like channels in that you can specify multiple ones and mix and match providers and you just pick which connected provider will be used for a specific trigger or flow action. We may not need to do that on v1.
 * Access to the classified intents in the flow is neat for sure, pretty sure that fits in nicely with the new engine's model around webhook results @rowanseymour what do you think on that front?
 * No need to call out NLU triggers in the flow dialog, ignoring triggers is ignoring triggers, regardless of type.

In any case, very excited about the thought that's gone into this. Unfortunately a lot of the implementation on the RapidPro side will have to be redone to fit into goflow but the great work you have done in thinking through UI implications are great.

PS. Was poking around the repo and the links to NLP side of things seems broken. Take it the NLP side is https://github.com/explosion/spaCy ?

**Comment by rowanseymour** [comment]
Exciting stuff! Thinking about how this would look in the new engine...

* There could be a "NLP provider" configuration passed in via the environment (this is where org specific stuff goes in the new world)
* Or NLP providers could be modeled as assets like resthooks and channels - one possible concern here is having to include credentials in assets 
* Calling an NLP provider has a certain amount of overhead.. and potential for failure.. so it feels like it warrants its own action, i.e. it can't just be a rule that makes a webhook call behind the scenes. 
* We'd probably want to model it as a "classify" node - which is a combination of ..
  - A classify action that..
     1. Calls out to the NLP provider
     2. Saves the classification as a result (whether it succeeded, then intents/entities as extra in the result (extra is where we store things additional data from router tests - e.g. like responses for webhook results)
  - A router made of rules that test the previously created result (in new world, rulesets are routers, and router tests are just expressions), e.g.
    * has_intent(results.foo, "register", 90)
    * has_entity(results.foo, "job", "farmer")
     
As Nic said - we are hopefully weeks away from switching to a completely new flow engine, so now is not the time to be adding functionality to it - but we can absolutely start discussing what this should look like in the new engine, and what you've done so far is a great proof of concept.

**Comment by nicpottier** [comment]
Hi all. Don't want this to die on the vine and I think there are ways we can begin making progress on this even if we can't put it all away until new editor.

Specifically, I think we can start talking about how we are going to model different NLP providers on an org and we can even start implementing the connection of those NLP providers.

@johncordeiro is that something you guys are ok leading? Would recommend using this issue as a way to talk through what models we see creating and then doing a PR based on that to connect two typer of NLP providers. (suggest wit.ai and bothub as a start)

Would only be seen by say people with beta flag enabled but will start this ball rolling and we can start thinking about how this plugs into new engine etc as well.

**Comment by johncordeiro** [comment]
@nicpottier, we're starting the development for this, I'll share the user story mapping for community approval next week.

**Comment by nicpottier** [comment]
@johncordeiro Please do not start any work until you chat with us (here) about it. For one I think for 5.2 the only supported NLP action will be classifying a message. (not triggers yet) The piece that would likely make the most sense for you guys to help out on is the connection of NLP providers to RapidPro and we should chat here on the proposed design of that. (likely something that looks vaguely channel like)

**Comment by smn** [comment]
Fwiw, [we've](https://turn.io) been using [Rasa](http://rasa.com/) and [Feersum](https://feersum.io/) for adding NLU in various ways in RapidPro for classification & intent. Almost all using some combination of RP webhook & serverless function to glue APIs together.

Happy to collaborate on some kind of pluggable provider for either of these if useful.

**Comment by nicpottier** [comment]
Thanks Simon. That only validates we probably need to address these as pluggable items like channels. It looks like Rasa has an endpoint for classifying:
   https://rasa.com/docs/rasa/nlu/using-nlu-only/

But not a lot of details on parameters.. do you know where they have some more docs?

What do your flows look like when you are doing these? Basically a call to the webhook then split on the classification results? What patterns have you seen emerge?

**Comment by johncordeiro** [comment]
[Rasa](http://rasa.com/), [Bothub](https://bothub.it) and Wit.ai (https://wit.ai) works in a similar way, you send a text message and as a response, the metadata for that giving intents and entities (information retrieved from the text). 

We started using Webhook calls and splitting the classification results and it was really tricky, especially because it's really common the need to classify responses in different NLP models to be more accurate, asides that are also necessary to verify the prediction confidence and create a threshold for that, so imagine how many `split by expression` and `webhook calls` nodes are necessary to make one simple accurate prediction inside the flow.

**Comment by nicpottier** [comment]
Right, ya, I'm saying getting exact examples across those platforms of request/response payload would help in figuring out the right abstraction. As you said, part of this is also figuring out what the right level of configuration for the call as well, IE, classify against model A on provider Z. (and model A only lives on provider Z)

So for each provider we will need a concept of which classifiers exist, have those synced etc.. but we need to see a few more examples to pick the right abstractions in RapidPro for those.

**Comment by smn** [comment]
For Rasa the docs are a bit hidden. They've recently launched Rasa X which is not open source but the Rasa NLU pieces still are and those are the only ones we're currently using. The HTTP API documentation for [Rasa NLU is here](https://rasa.com/docs/rasa/api/http-api/#tag/Model/paths/~1model~1parse/post).

What's interesting about Rasa's approach is that they have an emulation mode which allows the HTTP API to act like Wit (Facebook), Luis (Microsoft) or Dialogflow (Google) by supplying the `emulation_mode` query parameter. I'm not entirely sure how that prevents vendor lock in for Rasa itself but it perhaps does help when people have already invested in client implementations in those platforms and are wanting to continue seeing some returns on that investment. 

Most of these though follow the general pattern of allowing one to submit an expression and returning intents with a confidence score between zero and one and entities, if any, extracted. For many of the NLU models, being able to specify the language when attempting to parse an expression is fairly important as a big quality gain can be achieved with less training data if you are able to specify the language of the expression ahead of time.

Feersum's API is currently not documented publicly but it follows these same patterns.

Our approach has been similar to @johncordeiro, taking in an expression submitting it to a webhook, having it return the most likely intent with a confidence score. We generally have RapidPro determine whether or not the confidence score is high enough to continue with a certain path in the flow. The hassle here is that it adds a bunch of congnitive load to the people designing the flows and forces them to have a fairly deep understanding of some system that lives entirely outside of RapidPro. It's brittle and fairly difficult to debug as many of these NLU systems aren't deterministic. 

What I'm interested in is if RapidPro should at all have any opinion on how these NLU providers are trained or if that's a concern that lives entirely outside of RapidPro. What's important here is that it's not merely about submitting expressions for intent but also training on entities (which are generally substrings of the expressions) and being able to split these into testing and training samples and providing meaningful feedback on how well the model is performing. Tweaking the full expression set and resubmitting for training etc.

I think some kind of RapidPro Provider API that people can implement, perhaps similar to how channels work, would be a good start. As a minimum I'd expect to be able to submit an (expression, language) tuple for analysis, expect a list of (intent, confidence), and (entity, confidence) returned.

**Comment by johncordeiro** [comment]
The first version of the integration we've built to RapidPro works in this way you proposed, we both supported Bothub and Wit.ai. Just one example of each response:

Bothub:
```
{
  "intent": {
    "name": "love",
    "confidence": 0.6943462393863934
  },
  "intent_ranking": [
    {
      "name": "love",
      "confidence": 0.6943462393863934
    },
    {
      "name": "hate",
      "confidence": 0.30565376061360666
    }
  ],
  "labels_list": [
    "animal"
  ],
  "entities_list": [],
  "entities": {
    "animal": [
      {
        "value": "puppy",
        "entity": "dog",
        "confidence": 0.67255946125065845
      }
    ],
    "other": []
  },
  "text": "i love my puppy",
  "update_id": 47,
  "language": "en"
}
```

Wit.ai:
```
{
    "msg_id" : "dd526480-5784-43fa-a0f9-a3587fadc95d",
    "_text" : "temperature",
    "stats" : {
      "total_time" : 185
    },
    "entities" : {
      "intent" : [ {
        "confidence" : 0.6081958281101719,
        "value" : "temperature_get"
      }, {
        "confidence" : 0.3918041718898281,
        "value" : "temperature_set"
      } ]
    }
  }
```

In our experience, we could generify the usage in RapidPro flows very well on these both platforms, as @smn said, they follow a similar structure but the configuration side can be different. It sounds that Channel-like connections are a great start.

I can propose a UI/UX user story mapping for that, it would be the first step. Once both of you agree with the approach we can move forward.

**Comment by nicpottier** [comment]
@johncordeiro can you point us to your fork with those changes out of curiosity? Would give us a starting point to comment on approaches.

> What I'm interested in is if RapidPro should at all have any opinion on how these NLU providers are trained or if that's a concern that lives entirely outside of RapidPro.

I think at least for the first go that lands outside RapidPro. If we see a lot of usage and then can come up with a generic way that is an actual win for users then maybe that's something worth considering but I think the bar would have to be pretty high.

**Comment by johncordeiro** [comment]
@nicpottier we've added a [BaseConsumer](https://github.com/Ilhasoft/rapidpro/blob/34912dd5c6ce216edbe9d319b0e57171ea0e3d87/temba/nlu/models.py#L29) with basically some functions to `predict`, `get_intents `, `get_entities` and to validate the authorization token of the NLP Dataset in both platforms.

**Comment by johncordeiro** [comment]
<img height="500" alt="Screen Shot 2019-07-17 at 2 26 41 PM" src="https://user-images.githubusercontent.com/6073075/61403347-8d990300-a8ab-11e9-9fb7-6b0cf0a90b25.jpeg">

That's the User Story Mapping for Natural Language Understanding (NLU) capability in RapidPro. First of all, we've thought the feature for the common user segment of RapidPro, which are communication managers. NLU in RapidPro combines the power of conversation flows with AI-based predictions that guides it.

As we've experienced during the time we use this approach that the combination of both technologies builds the most powerful tool in the market for Chatbots with the flexibility and accessibility that it requires. We've benchmarked a lot of them but nothing compares to this.

## User Activities

Four main user activities were identified during its journey in the usage of NLU in RapidPro:

1. **Connect NLP Datasets**: Setup Natural Language Processing (NLP) Datasets that will provide predictions for RapidPro usage. As discussed, there are many NLP platforms in the market that can deliver predictions to usage in Chatbots. They follow a similar structure for predictions which is using HTTP REST API to provide `intent classification` (let's call IC from this point on) or `entity extraction` (let's call EX). Examples:

   1. [Bothub](https://bothub.it): An [open-source](http://github.com/ilhasoft/bothub) Software as a Service (SaaS) created on top of Rasa NLU and spaCy that supports 29 languages and several utilities for NLP Datasets as translations and evaluation module. Bothub basically retrieves `intent` node in its results that makes the `IC` process and the `entities`  list of features extracted from the sentence by doing the `EX` process. You can make realtime testing by using the following [dataset](https://bothub.it/johncordeiro/binary/Integration) or looking at the documentation on the same page.
   2. [Wit.ai](https://wit.ai/docs/http/20170307#get__message_link): SaaS acquired by Facebook some years ago that uses `entities` nomenclature for both `IC` and `EX` processes by just changing the type of entity retrieved. For `IC` it's used _trait_ entities and for `EX` are used _free-text_ and/or _keywords_. All of them are retrieved from `entities` list in the API result (as exposed [here](https://wit.ai/docs/http/20170307#get__message_link)), but its usage in RapidPro needs to be different. Our idea to converge its usage along with other platforms is given by setting up what entities are of the _trait_ kind (the API doesn't provide this info) for each wit.ai dataset added.
   3. [Rasa NLU]([https://rasa.com](https://rasa.com/)): Rasa NLU is an open-source software that's widely used by the Chatbot world, and its use a similar approach that Bothub does, by giving an `intent` node and `entities` list applying `IC` and `EX` respectively. Documentation for the API can be found [here](https://rasa.com/docs/rasa/api/http-api/#tag/Model/paths/~1model~1parse/post).

   Another important class that's commonly used by these platforms is the _model_ (also called _NLP Dataset_), which is a contextual content created by the user that at the end will give predictions (`IC` and `EX`) based on the training records added to it.

   As exposed it seems a very reasonable decision to use a Channel-like approach to add multiple `NLP Datasets` from different platforms, and converging its results in the structure of `IC` and `EX`.

2. **Consume NLP Datasets in Flows**: The consumption of `NLP Datasets` that are added to the RapidPro workspace can be primarily done by allowing users to categorize contact responses not just by `has any of the words` but also by `has an intent` for `IC` results and `has an entity` for `EX` ones. In the second section I'll expose in detail.

3. **Consume NLP Datasets in Triggers**: Another useful way of consuming `NLP Datasets` are using triggers. The human conversation follows a tree that each interaction of question/answer has a context involved. However, it's very common to change the subject during the conversation, especially if we're talking to a program, institution or company (as we usually do with RapidPro). 

   That's why the usage of triggers based on Natural Language Understanding is so powerful, it creates what no other visual platform like RapidPro can do: a Cognitive Tree. The details of this activity will be shared in the second section.

4. **Analyze Prediction Logs**: Completing the cycle, the recent prediction logs need to be exposed at the `NLP Dataset` page, displaying just simple information as normally is done by the webhook and channel logs.

## Steps and Details

The idea here is to detail and prototype each activity to get your approval and move forward with the coding phase.

### Connect NLP Datasets

#### Add NLP Dataset

The option to `Add NLP Dataset` can be in the same dropdown as `Add Channel`:

![Add NLU Provider](https://user-images.githubusercontent.com/6073075/61403480-c6d17300-a8ab-11e9-97dc-52c143b75e72.png)

#### Select NLP Platform

User needs to select one of the available NLP Providers (ignore non-monochromatic colors for now):
![Select NLP Platform](https://user-images.githubusercontent.com/6073075/61403521-d6e95280-a8ab-11e9-94e3-1523886b0500.png)


#### Manage NLP Datasets

The capability of removing an `NLP Dataset` that was added to the workspace is essential in this case, wherever this dataset is used in RapidPro it will stop validating.

### Consume NLP Datasets in Flows

Once the user had added at least one `NLP Dataset` in its workspace, new validation options are available in the `Wait for the contact to respond` node in RapidPro flows to apply results from `IC` (intent classification) and `EX` (entity extraction).

#### Set up rules with NLP Datasets available

- _has any of the intents_: select one of the datasets available to make `IC` and categorize user message based on Intents, which sends a request to the correspondent NLP platform added to the workspace. For this rule, a new field needs to be filled out, which is threshold confidence.
- _has any of the entities_: similarly to intent classification, we will use `EX` results to categorize.

Some assumptions that we want to follow on this step:

1. For the sake of efficiency, each user message will just send one request per `NLP Dataset` related to the ruleset, to avoid ambiguous requests.

2. The dataset will be displayed as a dropdown as exposed in the current approach we're doing today.

3. After each validation rule we can use `@results.response_1.extra` with the following payload sample:

   ```json
   {
   	"intent": "classA",
   	"entities": {
   		"entityA": "value",
   		"entityB": "value"
   	}
   }
   ```

4. The contact language will be sent as parameter to the NLP Platform (when available), as in the case of [Bothub](https://bothub.it) that one dataset can support different languages, so it will switch automatically based on this param.

### Consume NLP Datasets in Triggers

Once the user had added at least one `NLP Dataset` in its workspace, a new trigger option is activated and users can start flows based on `IC` (intent classification) results.

#### Setup NLP-based triggers

The main difference between Triggers approaches and Flows one is that trigger is capable of starting flows based on `IC` process only. As exposed before, the user experience for triggers is even more straightforward than inside a conversation, because contacts can simply change from one subject to another just making a new question about it.

Some assumptions that we want to follow on this step:

1. Setup threshold is also a requirement here, even because you just want to change the subject and start another flow if you have one-hundred percent sure that contacts really want it.

2. Filtering by Contact Groups is also a good thing we've experienced, allowing to create phases in conversation tree (which is also a good approach broadly discussed in the Chatbot field), for example, if you just want to answer questions of those ones who completed the Registration flow.

3. Setup priority among different NLP triggers helps to guide the conversation better when more than one intent was identified by multiple `NLP Datasets`. Let's say that we have a dataset to answer questions about greetings as `Hello` or `Good morning`, and the main purpose of your Chatbot is giving information about your company. So you will receive messages like this:

   - <u>User Message</u>: Hello, where are you based?
   - <u>Expected response</u>: We're based in the US.
   - <u>Greetings prediction</u>: 80% confidence for `Hello`
   - <u>Information prediction</u>: 78% confidence for `Address question`

   In this case, we want to answer the question instead of just answering `Hello!`, so `Address question` needs to have more priority. We could use a simple slider for priority selection or just create 5 groups of priority as `Lowest`, `Low`, `Medium`, `High` and `Highest`. If you agree with this approach my UX team can think the best way to display it.

4. Use the contact language to select the right one that dataset will predict.

5. Send the prediction results that triggered the flow in `@trigger.params`, with the following payload:

   ```json
   {
   	"intent": "classA",
   	"entities": {
   		"entityA": "value",
   		"entityB": "value"
   	}
   }
   ```

### Analyze prediction logs

Following RapidPro guidelines, users need to debug the results extracted from predictions as is normally done by Channels and Webhooks logs.

#### Open NLP Datasets Logs page

Similarly to Webhook logs, the prediction logs need to expose just recent predictions and the payload retrieved by the `NLP dataset`.

Some assumptions that we want to follow on this step:

1. Navigate through a list of predictions showing simple data and possibly viewing JSON results when clicking on it. For example:

<img width="795" alt="Screen Shot 2019-07-17 at 2 26 41 PM" src="https://user-images.githubusercontent.com/6073075/61403605-08fab480-a8ac-11e9-8545-b76c0338dfd2.png">

1. View failure stats when no intent was predicted because of an error in the request to the `NLP dataset`.
2. Create a simple chart (as Channel stats) that shows simply what intents are more demanded by contacts with counters.

------

That said, I would appreciate feedback on these points. We should validate the idea and make sure we have a clear understanding of all the details to start coding anything. We're willing to create a visual prototype and share with you guys as well.

**Comment by nicpottier** [comment]
Nice, this is super helpful and educational.

As you started, let's just formalize some terms so it is easier to chat about things (referenced this as well: https://developers.google.com/machine-learning/glossary/):
 * `Model` or `NLP Model` instead of `NLP Dataset`, just because Model seems more correct to start out with and because Dataset is really overloaded. Model is a single word and has a clearer meaning IMO. Some collision with Django, but I don't think that'll be too painful and our bias should be making things clearest for users.
 * `Classifier` instead of `Intent Classifier`, the latter is more precise but I think simplifying the language as much as possible for users is important. It then also makes the flow action easy `classify`
 * `Extractor` instead of `Entity Extractor` for the same reason. Feel less strongly about this, but just think `entity` may be a bit much for most users.
 * `Provider` instead of `Platform` just because it is a bit more generic and may fit into ideas we have around modeling airtime etc, as well. NLP may be self hosted so platform doesn't feel right.

Ok, anyways, will use that vocabulary below.

So important piece is we are not going to launch or develop this all at once. I think particularly around triggers there are some important questions about how that should work that I think will be partly informed by seeing how these services scale. I have some real concerns about any of these services being happy about getting millions of requests from us as we try to classify every incoming message. I do think it is interesting but let's not start there.

Let's focus on a flow categorization action first and try for the simplest thing that can work to start. For that I see the below high level tasks needed, each of these should get (at least) one new issue created so we can coordinate on just those pieces:

 1. Create pipeline to add new Model to an org. This would likely be something as you have above: pick provider, enter credentials, pick model. This then becomes an asset in the DB with provider specific config. (similar to current Channel) I think there are still some open questions around say, syncing of possible categories for a model that are important to answer. Ihlasoft can lead on this but we should all chat about details in a related issue.
 2. Create flow action in goflow, this likely uses the Provider abstraction we've been starting to talk about here (https://github.com/rapidpro/rapidpro/issues/997) but details have to be worked out. @rowanseymour can lead
 3. Create "Classify" action in floweditor that builds appropriate flow definition. @ericnewcomer can lead.
 4. Tie things together in mailroom, that'll be the implementation of the actual calls to the Providers that goflow will call. @nicpottier and Ihlasoft can lead.
 5. Ability to view provider logs in RapidPro. Ihlasoft can lead.

We can use this issue as a top level tracker of the sub-issues, but to keep things sane we should break these into smaller pieces referencing this issue.

Again, important to do this little by little, so after above is out and we get a feel for how everything is working in the field then we can add entity extraction and then start chatting about triggers. (I think at the very least this should be a single model an account can configure to classify responses with thresholds on particular categories that determine whether the trigger will fire but we can hash that out when we get there)

**Comment by johncordeiro** [comment]
Agreed with the terms. About your prioritization suggestion, we could consider just accepting `Extractor` in 5.2 release so we could validate the complete functionality inside flows before moving to triggers. 

About timelines, I'll talk with the team and plan the Sprint beginning date and share with you.

**Comment by IanLawrence** [comment]
> We generally have RapidPro determine whether or not the confidence score is high enough to continue with a certain path in the flow.

You mean have a user feedback flow 'What did you really mean?' which only tiggers if the confidence score is below a threshold? If so does this feed back to the model?

**Comment by smn** [comment]
@IanLawrence no, I meant, in RP we have some conditional branching where we determine if a confidence score for some intent is higher than a preconfigured acceptable threshold.

**Comment by johncordeiro** [comment]
@IanLawrence @smn @nicpottier I've been thinking about the trigger functionality, not specifically for NLP usage, but in general. What do you think if we have a trigger type named `Rule-based Trigger` that would enable users to define a rule that launches a flow, and these rules would be very similar to `Wait for Message` ruleset:

![Screen Shot 2019-08-20 at 5 02 16 PM](https://user-images.githubusercontent.com/6073075/63380167-4c918400-c36c-11e9-903b-1465dd9e83f7.png)

In this way, we could prioritize some rules on top of others and use `Classifier` and `Extractor` as we would normally do in flows. Let me know your thoughts about that ans we can open a new thread to discuss specifically about that.

**Comment by nicpottier** [comment]
Hah, too funny, I just wrote something about this very thing. 

Issue here: https://github.com/rapidpro/rapidpro/issues/1023

**Comment by nicpottier** [comment]
@johncordeiro any update on getting the Django side of NLP providers done? We are six weeks out from code freeze for 5.2 so we want to get moving on this. If your team can't do that side of things that's fine we can pick it up.

**Comment by johncordeiro** [comment]
we've started working on this, @matmsa27, that's conducting the work, will give you a completion date.

**Comment by matmsa27** [comment]
Hi @nicpottier , my name is Matheus and I work at Ilhasoft. It's a pleasure works on RapidPro and with you. Now I'm conduct the works for RapidPro at Ilhasoft. I working on this feature, and I believe that we need 3 weeks for done.

**Comment by nicpottier** [comment]
Hi @matmsa27, pleased to meet you. Can you outline a bit for me what you will be delivering in three weeks just to make sure we are on the same page?

**Comment by matmsa27** [comment]
Hi @nicpottier , I'm on the process transition at the Ilhasoft. Now we have a Platform team for work specially on RapidPro. On this transition process we are create a backlog, and we are studying and understanding the code. For start the feature, on the second week we will work on the tasks and description created [here on this comment](https://github.com/rapidpro/rapidpro/issues/832#issuecomment-512581458). On the first item assigned to us Ilhasoft, we will Create a pipeline to add a new Model to an Org (1.). After that on the third week, we will finalize the tests and implement the actual calls to the providers that goflow will call (4.), and then ability to view provider logs in RapidPro (5.).

**Comment by rowanseymour** [comment]
@johncordeiro @matmsa27 think this all live and working now. Can you guys at least test a bothub repository (classifier in the RP nomenclature) and verify that it creates intents and entities that you would expect? It's all on textit.in if you want to try it there or you can pull latest version of RP and mailroom if you have your own instance for testing.

**Comment by nicpottier** [comment]
Shipped, closing.
